### PR TITLE
queue remove: can remove successed,failed tasks msgs

### PR DESCRIPTION
**Pull request by karajan1001**
Followed from [#7592](https://github.com/iterative/dvc/issues/7592#issuecomment-1157176923) 

1. Seperate `remove` method to a new file.
2. Add `queued`, `failed`, `processed` flags to remove.
3. Add new unit tests for `queue status --queue/fail/success`
4. Implement methods to remove done tasks.
5. add two new unit test for `celery_queue.remove`
6. bump into dvc-task version 0.0.13

* [ ] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

A demo for this PR.
[asciicast](https://asciinema.org/a/PCejdbblayS3hkBmiRZmRTAwB)

### CODE REVIEW

**Comment by karajan1001** [comment]
@pmrowla 
The current problem is that the queued experiment I prepared will be executed wrongly by some celery worker in the background, making the tests really unstable. Any ideas about it?

File: dvc/repo/experiments/queue/base.py [hunk 1]
```diff
@@ -131,15 +131,27 @@ def put(self, *args, **kwargs) -> QueueEntry:
     def get(self) -> QueueGetResult:
         """Pop and return the first item in the queue."""
 
-    def remove(self, revs: Collection[str]) -> List[str]:
+    def remove(
+        self,
+        revs: Collection[str],
+        _all: bool = False,
```
Comment (pmrowla) [inline dvc/repo/experiments/queue/base.py:137]:
```suggestion
        all_: bool = False,
```

When trying to avoid collisions with reserved keywords either use a trailing underscore or change the name to something else like `remove_all`.

Try not to use a leading underscore for this (like `_all`), since the leading underscore is usually used to indicate either that a var is supposed to be private or that it is intentionally unused.

File: tests/func/experiments/test_queue.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        dvc.experiments.run(
+            failed_exp_stage.addressing,
+            params=[f"foo={i+queue_length}"],
+            queue=True,
+            name=name,
+        )
+    dvc.experiments.run(run_all=True)
+    output = capsys.readouterr()
+    assert "Failed to reproduce experiment" in output.err
+    return name_list
```
Comment (pmrowla) [inline tests/func/experiments/test_queue.py:60]: Fixtures should not contain tests themselves, they should only provide any necessary setup and/or cleanup needing to run tests that use the fixture.

It's general it would be ok to have a test fixture that sets up some set of failed experiments, but `test_failed_exp_celery` should be separate from that. It does not make sense to have a repeated UI test that runs every time we re-use that fixture.

However, I think that you probably won't be able to use fixtures that run celery experiments in this way. We will end up hitting race conditions due to how the pytest-celery worker executes tasks.

Basically it will be safer to not use fixtures that run experiments, and setup everything in individual tests.

File: tests/func/experiments/test_queue.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    assert len(status) == 9
+    for task in failed_tasks:
+        assert status[task] == "Failed"
+    for task in success_tasks:
+        assert status[task] == "Success"
+    for task in queued_tasks:
+        assert status[task] == "Queued"
+
+
+def test_queue_remove(dvc, failed_tasks, success_tasks, queued_tasks):
```
Comment (pmrowla) [inline tests/func/experiments/test_queue.py:98]: We don't really need a func test for this since we already have a unit test for queue remove. The only additional thing the func test is doing is `status()` checks, but `status()` should be tested separately anyways.

File: tests/func/experiments/test_queue.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    queue = dvc.experiments.celery_queue
+    done_result = first(queue.iter_done())
+    name = done_result.entry.stash_rev
+    captured = capsys.readouterr()
+    queue.logs(name, follow=follow)
+    captured = capsys.readouterr()
+    assert "failed to reproduce 'failed-copy-file'" in captured.out
+
+
+def test_queue_status(dvc, failed_tasks, success_tasks, queued_tasks):
```
Comment (pmrowla) [inline tests/func/experiments/test_queue.py:85]: Rather than executing experiments, it would be better to test `status()` unit test `status()` with mocked results of `iter_queued/iter_active/iter_done`

**Review by pmrowla — CHANGES_REQUESTED** [review]

**Comment by pmrowla** [comment]
> @pmrowla The only problem remained is that the queued experiment I prepared will be executed wrongly by some celery worker in the background, making the tests really unstable. Any ideas about it?

@karajan1001 it's likely due to using the celery pytest worker inside multiple fixtures at once.

In general, we should try to move away from adding complex func tests in favor of unit testing whenever possible. In this case, for celery we do need a couple of func tests to make sure that execution and the UI work properly (in both failed and successful cases), but for testing specific backend or API calls (like remove/status) it's better to use unit tests + mocked celery behavior

**Comment by pmrowla** [comment]
One other thing to note here regarding commits, ideally we want to keep changes in separate commits where possible. In this PR it would be nice to have the separation with:

- 1 commit for the dependency version bump. If there are specific changes required to make the new dep version work with DVC then those changes can also go into the version bump commit (like when a dependency has a backwards incompatible API change)
- 1 commit for the `queue remove` related changes
- 1 commit for the `queue status` related changes

Basically for small "review fix" type changes it makes sense to squash them into the larger commit, but for larger independent changes that affect different parts of the code base or different parts of the CLI/API, having separate commits makes it easier to keep track of what changed and when. So follow-up changes/fixes to `remove` or `status` can get squashed into each relevant commit before merging the PR, but ideally the final PR would not contain commits that mix changes to both `remove` and `status`.

---

This isn't really a requirement for this PR, but it's something to be aware of in the future

**Comment by karajan1001** [comment]
> One other thing to note here regarding commits, ideally we want to keep changes in separate commits where possible. In this PR it would be nice to have the separation with:
> 
> * 1 commit for the dependency version bump. If there are specific changes required to make the new dep version work with DVC then those changes can also go into the version bump commit (like when a dependency has a backwards incompatible API change)
> * 1 commit for the `queue remove` related changes
> * 1 commit for the `queue status` related changes

After removing the functional test of `queue status`. It only contains `queue remove` related changes and the dependency bump. I can separate them later.

File: dvc/repo/experiments/queue/base.py (hunk 1 above)
Comment (pmrowla) [inline dvc/repo/experiments/queue/base.py:137]:
```suggestion
        all_: bool = False,
```

looks like this one was missed

File: dvc/repo/experiments/queue/remove.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        if success:
+            done_tasks.extend(self.iter_success())
+        done_entries = [result.entry for result in done_tasks]
+        _remove_done_tasks(self, done_entries)
+        removed.extend(_get_names(done_tasks))
+
+    return removed
+
+
+def remove(
```
Comment (pmrowla) [inline dvc/repo/experiments/queue/remove.py:128]:
```suggestion
def celery_remove(
```

Since this is in `queue/remove.py` and is specific to the Celery version of the queue, make sure to name it appropriately so it's obvious this is not used in the base queue or other local queue types

File: dvc/repo/experiments/queue/remove.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    all_: bool = False,
+) -> List[str]:
+    """Remove the specified entries from the queue.
+
+    Arguments:
+        revs: Stash revisions or queued exp names to be removed.
+        queued: Remove all queued tasks.
+        failed: Remove all failed tasks.
+        success: Remove all success tasks.
+        all: Remove all tasks.
```
Comment (pmrowla) [inline dvc/repo/experiments/queue/remove.py:143]:
```suggestion
        all_: Remove all tasks.
```

File: dvc/repo/experiments/queue/remove.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                names.append(entry.result.ref_info.name)
+                continue
+            entry = entry.entry
+        name = entry.name
+        name = name or entry.stash_rev[:7]
+        names.append(name)
+    return names
+
+
+def clear(self: "LocalCeleryQueue", **kwargs) -> List[str]:
```
Comment (pmrowla) [inline dvc/repo/experiments/queue/remove.py:95]:
```suggestion
def celery_clear(self: "LocalCeleryQueue", **kwargs) -> List[str]:
```

File: dvc/repo/experiments/queue/base.py [hunk 8]
```diff
@@ -213,12 +208,17 @@ def _format_entry(
...
+        )
+        result.extend(
+            _format_entry(queue_entry, status="Success")
+            for queue_entry, _ in self.iter_success()
         )
         return result
 
-    def _remove_revs(self, stash_revs: Mapping[str, ExpStashEntry]):
+    @staticmethod
+    def _remove_revs(stash_revs: Mapping[str, ExpStashEntry], stash: ExpStash):
```
Comment (pmrowla) [inline dvc/repo/experiments/queue/base.py:221]: Since this is no longer really a private method it should be moved out of the queue classes and into `experiments/stash.py` as `ExpStash.remove_revs(self, stash_revs)`

File: tests/func/experiments/test_queue.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    queue = dvc.experiments.celery_queue
+    done_result = first(queue.iter_done())
+    name = done_result.entry.stash_rev
+    captured = capsys.readouterr()
+    queue.logs(name, follow=follow)
+    captured = capsys.readouterr()
+    assert "failed to reproduce 'failed-copy-file'" in captured.out
+
+
+def test_queue_remove_done(dvc, failed_tasks, success_tasks):
```
Comment (pmrowla) [inline tests/func/experiments/test_queue.py:67]: Do we need still need this test? With unit tests for `remove()` we should not need any additional func tests here

**Review by pmrowla — CHANGES_REQUESTED** [review]

File: tests/func/experiments/test_queue.py (hunk 9 above)
Comment (karajan1001) [reply tests/func/experiments/test_queue.py:67]: I think we can simplify them. We tested the remove method case by case as well as the cli test. But we don't have a place to test the whole process. This is why I added a functional test here, But I think we can simplify it, We don't need to test all of the cases, but only one of them.

**Review by pmrowla — APPROVED** [review]
